docs/packaging-your-sql-project/intro-to-etl-pipelines-with-python-and-sql.md

```diff
@@ -11,4 +11,186 @@ kernelspec:
+
+2. `pandas` is an extensive package. Comprehensively going through the entire library's documentation would be time-consuming and unnecessary for your understanding of the script. The script relies only on typical `pandas` functions, such as ones that help rename columns appropriately, clean data for null values, and create new columns. 
+
+Let's now dissect `datadownload.py` and focus specifically on key lines of its ETL process.
+
+![](etl-co2-sample.jpg)
+
+### Extracting with Python
+
+Take a look at the function `fuel_consumption_metadata_extraction`. On lines 100 - 115, we access the Government of Canada's API and extract data regarding fuel consumption ratings that are in English. The output of this function is a `pandas` data frame with two columns: "name" and "url". Each row in the data frame corresponds to a single data set under the fuel consumptions ratings portfolio that is in English.  
+
+```python
+url_open_canada = "https://open.canada.ca/data/api/action/package_show?id=98f1a129-f628-4ce4-b24d-6f16bf24dd64"  # noqa E501
+json_resp = requests.get(url_open_canada)
+# Check response is successful and application is of type JSON
+if (
+    json_resp.status_code == 200
+    and "application/json" in json_resp.headers.get("Content-Type", "")
+):
+    # Format data and obtain entries in english
+    open_canada_data = json_resp.json()
+    data_entries = pd.json_normalize(
+        open_canada_data["result"], record_path="resources"
+    )
+    data_entries["language"] = data_entries["language"].apply(
+        lambda col: col[0]
+    )
+    data_entries_english = data_entries[
+        data_entries["language"] == "en"
+    ]  # noqa E501
+    final_result = data_entries_english[["name", "url"]]
+```
+
+We then finalize the extraction process by saving each data set in its raw format as its own `pandas` data frame from  `final_result`. This is done with the `extract_raw_data` functions under a `for loop` in the script's main call on lines 342 - 358.
+
+```python
+# Fuel consumption metadata extraction urls
+data_entries_english = fuel_consumption_metadata_extraction()
+
+for item in data_entries_english.iterrows():
+    name, url = item[1]["name"], item[1]["url"]
+
+    if "Original" in name:
+        continue
+
+    # Form file name
+    file_name = f'{name.replace(" ","_")}.csv'
+
+    # Extract raw data
+    item_based_url = extract_raw_data(url)
+
```

Comment:
A suggestion is instead of showing the main program, that you show key functions in the script

the main program snippets feel really out of place and confusing

Reply:
Main program snippets have been replaced with function explanations.